Allow editing an existing Signing Request & generate new CSR

sam-bertin opened the pull request:
Closes https://github.com/maykinmedia/django-simple-certmanager/issues/41

- Regenerate the CSR / PK from the Subject Information
- Added a boolean field to trigger the regeneration of the CSR
- Updated the SigningRequest model to make the CSR and PK fields editable.
- Admin_client tests

sam-bertin committed:
[#41] :children_crossing: Editing an existing Signing Request

Related to Taiga internal 718

- Regenerate the CSR / PK from the Subject Information
- Added a boolean field to trigger the regeneration of the CSR
- Updated the SigningRequest model to make the CSR and PK fields editable.
- Admin_client tests

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -82,11 +56,51 @@ class SigningRequestAdmin(admin.ModelAdmin):
     readonly_fields = ("csr",)
     actions = [download_csr]
 
+    def get_fieldsets(self, request, obj=None):
```
I think you can do revert this and instead in the model form check if an instance is provided or not:

```py
def __init__(self, *args, **kwargs):
    super().__init__(*args,**kwargs)
    if not self.instance.pk:  # adding a new object instead of updating an existing one
        del self.fields["should_renew_csr"]
```

edit: I see that's what you're doing already. I think you can safely include the field in the static `fieldsets` and django is smart enough to not crash.

sam-bertin replied:
If I hardcode `should_renew_csr` in the fieldset than it gives me a KeyError when creating a `SigningRequest`

![image](https://github.com/user-attachments/assets/150f5c88-f331-4cf2-93ab-bc437668a4f4)

I don't see an alternative to dynamically generating the fieldset.

sergei-maertens replied:
you're right, I thought the admin was a bit smarter than that.

Since I found this quite ugly to have to make the fieldsets dynamic, instead I've opted to toggle the disabled state from the checkbox, so the field is now always included, but can only be edited for existing instances.

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -2,7 +2,9 @@
 from django.utils.translation import gettext_lazy as _
 
-from .models import Certificate
+from simple_certmanager.csr_generation import generate_private_key_with_csr
```
```suggestion
from .csr_generation import generate_private_key_with_csr
```

sergei-maertens commented on simple_certmanager/forms.py:
```diff
@@ -13,6 +15,43 @@
+        if not self.instance.pk:
+            self.fields.pop("should_renew_csr")
+
+    def clean_should_renew_csr(self):
+        # Regenerate the CSR/Private Key if the checkbox is checked
```
you really have to see validation and processing as two separate things - first everything must be validated before you can process based on the input data.

This implementation would persist changes while other fields may be invalid, so then you get validation errors but also changes will be applied and that's mega confusing.

During validation, you may *only* validate. Do not touch `self.instance`, and especially don't write to the database or filesystem. Instead, validate everything, and once validation passes, you process the changes in `def save`.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -39,13 +39,11 @@ class SigningRequest(models.Model):
         help_text=_("Email address for the certificate"), blank=True
     )
     csr = models.TextField(
-        editable=False,
```
there is no reason to make these fields editable and expose them through forms, we manage the values programmatically

sam-bertin replied:
Oh yeah I confused editable for all types of edits (since we now allow the edition of the CSR/PK values).

sergei-maertens commented:
Additionally, can you mention the github issue ID in the commit message instead of the Taiga issue? That way we can click through from the commit to the issue and vice versa.

sergei-maertens committed:
:art: [#41] Use fixed fieldsets but make form field disabled

* Added a test to show the problematic behaviour when the field is
  just simply deleted from the form.
* Instead of deleting the field from the form, make it disabled. The
  outcome is the same - no value for the checkbox is sent, so the
  boolean field is always interpreted as False, which, in this case,
  means: do nothing.

sergei-maertens commented:
I've rebased to squash your two commits together since they logically are a single unit, and fixed up the issue references so that things properly cross-link in Github.

I've also added a test to demonstrate the key error you run into, since the existing tests didn't reveal that problem.
